### PR TITLE
respect the migration level in the backup

### PR DESCRIPTION
**Pull request by MaximilianMeister**
backup the migration level in the meta.yml, and restore it before loading the database

### CODE REVIEW

File: crowbar_framework/app/models/backup.rb [hunk 1]
```diff
@@ -167,6 +171,7 @@ def save_archive
     self.version = meta["version"]
     self.size = path.size
     self.created_at = Time.zone.parse(meta["created_at"])
+    self.migration_level = meta["migration_level"]
```
Comment (vuntz) [inline crowbar_framework/app/models/backup.rb]: Please add a fallback: this patch won't be in the release, so we need to assume there'll be backups without this metadata.

Comment (rsalevsky) [inline crowbar_framework/app/models/backup.rb]: Ouch, this means that we take the risk that a backup restore / upgrade will fail.

Comment (vuntz) [inline crowbar_framework/app/models/backup.rb]: Why? We know what the fallback should be.

File: crowbar_framework/lib/crowbar/migrate.rb [hunk 2]
```diff
@@ -31,6 +31,14 @@ def migrate!
           )
         end
       end
+
+      def migrate_to(level)
+        puts "Migrating database schema to #{level}..."
+        ActiveRecord::Migrator.migrate(
+          ActiveRecord::Migrator.migrations_paths,
+          level.to_i
```
Comment (MaximilianMeister) [inline crowbar_framework/lib/crowbar/migrate.rb]: just to mention, that also seems to work for down migrations. so when a Cloud6 backup is older than the `current_version` it will migrate down

**Comment by MaximilianMeister** [comment]
@vuntz I have created a new migration now, instead of touching the one that is already frozen for the release

**Comment by tboerger** [comment]
:+1:

**Comment by rsalevsky** [comment]
+1

**Comment by vuntz** [comment]
Is it still untested? (see PR title)

**Comment by MaximilianMeister** [comment]
> Is it still untested? (see PR title)

yes it is, I will test it today later on

**Comment by MaximilianMeister** [comment]
>> Is it still untested? (see PR title)

> yes it is, I will test it today later on

tested and works. on the first migration it migrates to what it says in meta.yml, and on the second migration all the way up to the latest

File: crowbar_framework/lib/crowbar/backup/restore.rb [hunk 3]
```diff
@@ -296,8 +297,12 @@ def proposal?(file)
         raw_data.key?("attributes") && raw_data.key?("deployment")
       end
 
-      def migrate_database(time)
-        Crowbar::Migrate.migrate!
+      def migrate_database(time, migration_level = nil)
+        if migration_level
+          Crowbar::Migrate.migrate_to(migration_level)
```
Comment (vuntz) [inline crowbar_framework/lib/crowbar/backup/restore.rb]: What happens if that migration level is not known yet? (ie: use new backup on old code)

Comment (MaximilianMeister) [inline crowbar_framework/lib/crowbar/backup/restore.rb]: good point, but is this really possible in the real life? that implies that a user somehow has downgraded his cloud. or where would the user get that backup from?

Comment (vuntz) [inline crowbar_framework/lib/crowbar/backup/restore.rb]: > good point, but is this really possible in the real life? that implies that a user somehow has downgraded his cloud. or where would the user get that backup from?

If we had a db migration over the lifetime of the product, it could happen if the user does a backup with updates, and restores without updates.

Comment (MaximilianMeister) [inline crowbar_framework/lib/crowbar/backup/restore.rb]: > If we add a db migration over the lifetime of the product, it could happen if the user does a backup with updates, and restores without updates.

that's the point that I don't understand, why would he restore without the updates? wouldn't that mean that the updates were rolled back?

Comment (vuntz) [inline crowbar_framework/lib/crowbar/backup/restore.rb]: > that's the point that I don't understand, why would he restore without the updates?

Because a user does mistakes.

Comment (MaximilianMeister) [inline crowbar_framework/lib/crowbar/backup/restore.rb]: back to the question 

> What happens if that migration level is not known yet? (ie: use new backup on old code)

I think I we were not on the same page, I thought you mean the migration_level attribute, or were you talking about the migration_level in the db?

if it is about the attribute: 
wouldn't it just be ignored then? it depends on how old the code is I guess. but if it is at the code state of release, then the migration just runs two times without the migration_level distinction.

if it is about the database entry: 
we could just add a check if that migration exists in `migrator.get_all_versions` and when it doesn't exist we return and show a message in the UI.
would that be fine with you?

Comment (vuntz) [inline crowbar_framework/lib/crowbar/backup/restore.rb]: Let me rephrase my point: if the migration level attribute in the backup is higher than what the rails app knows, then I think we should either do not accept to upload this backup or fail early when trying to restore the backup.

I hope it helps? (I'm a bit confused by how you split the options, but that's probably an issue in my brain ;-))

Comment (MaximilianMeister) [inline crowbar_framework/lib/crowbar/backup/restore.rb]: > I hope it helps?

I hope https://github.com/MaximilianMeister/crowbar-core/commit/5ad4362e3ea165174fcac82b1888bf97638f5781 helps :-)

File: crowbar_framework/lib/crowbar/backup/restore.rb [hunk 4]
```diff
@@ -268,14 +269,23 @@ def run_installer
...
-        migrate_database(:before)
+        if ActiveRecord::Migrator.get_all_versions.include? @migration_level
+          migrate_database(:before, @migration_level)
+        else
+          Rails.logger.error "Cannot migrate to #{@migration_level}. Migration unknown"
+          set_failed
+          @status[:restore_database] = {
+            status: :not_acceptable,
+            msg: I18n.t("backups.index.restore_database_failed")
+          }
```
Comment (vuntz) [inline crowbar_framework/lib/crowbar/backup/restore.rb]: Should we `return` here?

File: crowbar_framework/lib/crowbar/backup/restore.rb (hunk 3 above)
Comment (vuntz) [inline crowbar_framework/lib/crowbar/backup/restore.rb]: Thanks, it does!

File: crowbar_framework/lib/crowbar/backup/restore.rb (hunk 4 above)
Comment (MaximilianMeister) [inline crowbar_framework/lib/crowbar/backup/restore.rb]: yes we should :-) thanks

Comment (vuntz) [inline crowbar_framework/lib/crowbar/backup/restore.rb]: Not related to your change here, but now this raises the question: should we return also after https://github.com/crowbar/crowbar-core/pull/343/files#diff-f70b9c61f1886e37f29a33a5988e642eR291 ? :-)

Comment (MaximilianMeister) [inline crowbar_framework/lib/crowbar/backup/restore.rb]: heh, I think this was forgotten when we introduced the second migrate_database call... thanks again

**Comment by vuntz** [comment]
Thanks for all the fixes, +1.

**Comment by rsalevsky** [comment]
+1

I guess some parts needs to go to cloud 5?

**Comment by MaximilianMeister** [comment]
normal gating is green. but I triggered rebuild with `plain crowbarbackup crowbarrestore testsetup`
https://ci.suse.de/job/openstack-mkcloud/17319/console

**Comment by MaximilianMeister** [comment]
https://github.com/crowbar/crowbar-core/pull/355 should fix the issue in the gating, so I guess we would have to wait for it to be merged

File: crowbar_framework/app/models/backup.rb [hunk 5]
```diff
@@ -172,6 +176,7 @@ def save_archive
     self.version = meta["version"]
     self.size = path.size
     self.created_at = Time.zone.parse(meta["created_at"])
+    self.migration_level = meta["migration_level"] || ActiveRecord::Migrator.last_migration.version
```
Comment (vuntz) [inline crowbar_framework/app/models/backup.rb]: `ActiveRecord::Migrator.last_migration.version` is wrong here; I think it should a hard-coded version which is the latest one we have as of today.

Comment (MaximilianMeister) [inline crowbar_framework/app/models/backup.rb]: >  I think it should a hard-coded version which is the latest one we have as of today.

so that we come to a state of cloud6 without updates? what would happen with that hard coded value when we add another migration?

Comment (vuntz) [inline crowbar_framework/app/models/backup.rb]: > so that we come to a state of cloud6 without updates?

(not sure what you mean)

> what would happen with that hard coded value when we add another migration?

Well, that's where the second call to migrate is used :-) This is what happens:

- option a: backup doesn't have this migration level metadata. It means that it was created before this PR got merged, which means that the migration level is the hardcoded one I suggest. (It cannot be lower than that due to the fact that the backup feature was non-existing before)

- option b: backup has this metadata. It means our first call to migrate moves us to this level, we do the restore, then we call migrate to the latest level.

I'm not sure if what I'm saying is clear. Does that make sense?

File: crowbar_framework/app/models/backup.rb [hunk 6]
```diff
@@ -172,6 +176,8 @@ def save_archive
     self.version = meta["version"]
     self.size = path.size
     self.created_at = Time.zone.parse(meta["created_at"])
+    # 20151222144602_create_backups.rb
```
Comment (MaximilianMeister) [inline crowbar_framework/app/models/backup.rb]: @vuntz to have some comment, can you help me here a bit to find a reasonable explanation for this in your words?

**Comment by MaximilianMeister** [comment]
gating will not make sense until https://github.com/crowbar/crowbar-core/pull/359 is addressed

**Comment by MaximilianMeister** [comment]
backup restore gating succeeded

File: crowbar_framework/db/migrate/20160215120522_add_migration_level_to_backups.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class AddMigrationLevelToBackups < ActiveRecord::Migration
+  def change
+    add_column :backups, :migration_level, :integer, limit: 8
```
Comment (vuntz) [inline crowbar_framework/db/migrate/20160215120522_add_migration_level_to_backups.rb]: Do we need to set somehow the default value for existing backups?

Comment (MaximilianMeister) [inline crowbar_framework/db/migrate/20160215120522_add_migration_level_to_backups.rb]: we could use `default: 20151222144602` like for the backup upload in https://github.com/crowbar/crowbar-core/pull/343/files#diff-408ac208d889893305de65518aa046a2R180 ?

Comment (MaximilianMeister) [inline crowbar_framework/db/migrate/20160215120522_add_migration_level_to_backups.rb]: this can also be changed in another migration if needed with `change_column_default` in the future

Comment (vuntz) [inline crowbar_framework/db/migrate/20160215120522_add_migration_level_to_backups.rb]: I'd use that as a default, yes.

**Comment by vuntz** [comment]
+1

**Comment by rsalevsky** [comment]
Still +1

**Comment by MaximilianMeister** [comment]
rebased

**Comment by vuntz** [comment]
Still +1

**Comment by MaximilianMeister** [comment]
rebased

**Comment by MaximilianMeister** [comment]
@vuntz can you merge this?

**Comment by vuntz** [comment]
> @vuntz can you merge this?

We're missing the +1 after the rebase.

Here's mine: +1.

**Comment by rsalevsky** [comment]
+1

**Comment by vuntz** [comment]
@MaximilianMeister feel free to backport to 3.0.

**Comment by MaximilianMeister** [comment]
> We're missing the +1 after the rebase.

@vuntz sorry for the nagging, but why do we need another +1 after a rebase?
